File: medium/3302-find-the-lexicographically-smallest-valid-sequence/solution.java
/**
 * 3302. Find the Lexicographically Smallest Valid Sequence
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public int[] validSequence(String word1, String word2) {
        int n = word1.length(), m = word2.length();

        // f[i] = # of trailing chars of word2 matchable exactly by word1[i:]
        int[] f = new int[n + 1];
        int k = 0;
        for (int i = n - 1; i >= 0; i--) {
            if (k < m && word1.charAt(i) == word2.charAt(m - 1 - k)) {
                k++;
            }
            f[i] = k;
        }

        int[] res = new int[m];
        int j = 0, i = 0;
        boolean changed = false;
        while (i < n && j < m) {
            if (word1.charAt(i) == word2.charAt(j)) {
                res[j++] = i;
                i++;
            } else if (!changed && f[i + 1] >= m - j - 1) {
                // spend the single change here; the rest still matches exactly
                res[j++] = i;
                i++;
                changed = true;
            } else {
                i++;
            }
        }

        return j == m ? res : new int[0];
    }
}

File: medium/3302-find-the-lexicographically-smallest-valid-sequence/solution.py
from typing import List


class Solution:
    def validSequence(self, word1: str, word2: str) -> List[int]:
        """
        3302. Find the Lexicographically Smallest Valid Sequence
        Time: O(n)
        Space: O(n)
        """
        n, m = len(word1), len(word2)

        # f[i] = # of trailing chars of word2 matchable exactly by word1[i:]
        f = [0] * (n + 1)
        k = 0
        for i in range(n - 1, -1, -1):
            if k < m and word1[i] == word2[m - 1 - k]:
                k += 1
            f[i] = k

        res = []
        j = 0
        changed = False
        i = 0
        while i < n and j < m:
            if word1[i] == word2[j]:
                res.append(i)
                j += 1
                i += 1
            elif not changed and f[i + 1] >= m - j - 1:
                # spend the single change here; the rest still matches exactly
                res.append(i)
                j += 1
                i += 1
                changed = True
            else:
                i += 1

        return res if j == m else []
